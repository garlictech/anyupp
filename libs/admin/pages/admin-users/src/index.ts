import { AdminPagesAdminUsersModule } from './lib/admin-pages-admin-users.module';
export { AdminUserFormService } from './lib/services/admin-user-form.service';

export { AdminPagesAdminUsersModule };
