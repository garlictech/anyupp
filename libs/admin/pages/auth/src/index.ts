import { AdminPagesAuthModule } from './lib/admin-pages-auth.module';

export { AdminPagesAuthModule };
