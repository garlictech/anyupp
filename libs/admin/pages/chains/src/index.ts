import { AdminPagesChainsModule } from './lib/admin-pages-chains.module';

export { ChainFormService } from './lib/services/chain-form.service';
export { AdminPagesChainsModule };
export * from './lib/pure/style-color-tools';
