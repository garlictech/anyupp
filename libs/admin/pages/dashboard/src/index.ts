import { AdminPagesDashboardModule } from './lib/admin-pages-dashboard.module';

export { AdminPagesDashboardModule };