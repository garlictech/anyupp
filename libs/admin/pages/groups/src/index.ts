import { AdminPagesGroupsModule } from './lib/admin-pages-groups.module';

export { GroupFormService } from './lib/services/group-form.service';
export { GroupListService } from './lib/services/group-list.service';
export { AdminPagesGroupsModule };
