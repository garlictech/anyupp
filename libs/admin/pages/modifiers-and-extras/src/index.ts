export * from './lib/admin-pages-modifiers-and-extras.module';
export { ProductComponentSetFormComponent } from './lib/components/product-component-set-form/product-component-set-form.component';
export { ModifiersAndExtrasFormService } from './lib/services/modifiers-and-extras-form.service';
export { ModifiersAndExtrasListService } from './lib/services/modifiers-and-extras-list.service';
