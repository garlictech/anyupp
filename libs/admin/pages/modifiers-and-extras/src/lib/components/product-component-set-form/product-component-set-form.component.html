<nb-card>
  <nb-card-header>
    <div class="float-start">
      {{
        (productComponentSet?.id
          ? 'productComponentSets.editProductComponentSet'
          : 'productComponentSets.newProductComponentSet'
        ) | translate
      }}
    </div>
    <button
      nbButton
      ghost
      shape="round"
      size="small"
      class="float-right"
      (click)="close()"
    >
      <nb-icon icon="close-outline"></nb-icon>
    </button>
  </nb-card-header>
  <nb-card-body>
    <form [formGroup]="dialogForm!" (ngSubmit)="submit()" autocomplete="off">
      <div class="row">
        <div class="col-md-4">
          <bgap-form-select
            caption="common.chain"
            [localize]="false"
            [options]="(chainOptions$ | async) || []"
            [control]="dialogForm.get('chainId')! | asFormControl"
            inputId="chainId"
          ></bgap-form-select>
        </div>
        <div class="col-md-4">
          <bgap-form-select
            caption="common.type"
            [localize]="false"
            [options]="typeOptions"
            [control]="dialogForm.get('type')! | asFormControl"
            inputId="type"
          ></bgap-form-select>
        </div>
      </div>

      <bgap-form-text-input
        caption="common.nameHU"
        type="text"
        [control]="dialogForm.get('name')!.get('hu')! | asFormControl"
        inputId="nameHu"
      ></bgap-form-text-input>
      <bgap-form-text-input
        caption="common.nameEN"
        type="text"
        [control]="dialogForm.get('name')!.get('en')! | asFormControl"
        inputId="nameEn"
      ></bgap-form-text-input>
      <bgap-form-text-input
        caption="common.nameDE"
        type="text"
        [control]="dialogForm.get('name')!.get('de')! | asFormControl"
        inputId="nameDe"
      ></bgap-form-text-input>

      <bgap-form-text-input
        caption="common.description"
        type="text"
        [control]="dialogForm.get('description')! | asFormControl"
        inputId="description"
      ></bgap-form-text-input>

      <bgap-form-serving-mode
        [control]="dialogForm.get('supportedServingModes')! | asFormControl"
        [servingModes]="servingModes"
      >
      </bgap-form-serving-mode>

      <fieldset class="form mb-4">
        <legend class="label">
          {{ 'productComponentSets.productComponents' | translate }}
        </legend>

        <div class="row">
          <div class="col-md-6">
            <bgap-form-select
              caption="productComponentSets.productComponent"
              [localize]="true"
              [options]="productComponentOptions"
              [control]="
                componentForm.get('productComponentId')! | asFormControl
              "
              inputId="productComponentId"
            ></bgap-form-select>

            <button
              nbButton
              size="small"
              type="button"
              status="primary"
              (click)="addComponentToList()"
              [disabled]="componentForm.invalid"
            >
              {{ 'common.add' | translate }}
            </button>
          </div>
          <div
            class="col-md-6"
            [hidden]="
              dialogForm.value?.type === eProductComponentSetType.MODIFIER
            "
          >
            <bgap-form-text-input
              caption="productComponentSets.maxSelection"
              type="number"
              [control]="dialogForm.get('maxSelection')! | asFormControl"
              inputId="maxSelection"
            ></bgap-form-text-input>
          </div>
        </div>

        <div class="mb-3 mt-3">
          <nb-list-item
            *ngFor="
              let componentId of dialogForm.value?.items || [];
              let i = index;
              first as isFirst;
              last as isLast
            "
          >
            <div class="w-100">
              <div class="float-start">
                <div class="subtitle-2">
                  {{ productComponentObject[componentId] | localize }}
                </div>
              </div>

              <button
                nbButton
                shape="round"
                status="danger"
                size="tiny"
                class="float-right me-2"
                type="button"
                [title]="'common.delete' | translate"
                (click)="removeComponentFromList(componentId)"
              >
                <nb-icon icon="trash-2-outline"></nb-icon>
              </button>

              <button
                [hidden]="isLast"
                nbButton
                shape="round"
                size="tiny"
                type="button"
                class="float-right me-2"
                [title]="'common.moveDown' | translate"
                (click)="move(i, 1)"
              >
                <nb-icon icon="chevron-down"></nb-icon>
              </button>
              <button
                [hidden]="isFirst"
                nbButton
                shape="round"
                size="tiny"
                type="button"
                class="float-right me-2"
                [title]="'common.moveUp' | translate"
                (click)="move(i, -1)"
              >
                <nb-icon icon="chevron-up"></nb-icon>
              </button>
            </div>
          </nb-list-item>
        </div>
      </fieldset>

      <div class="mt-6">
        <button
          nbButton
          size="small"
          type="submit"
          fullWidth
          status="primary"
          [disabled]="dialogForm.invalid"
        >
          {{ 'common.submit' | translate }}
        </button>
      </div>
    </form>
  </nb-card-body>
</nb-card>
