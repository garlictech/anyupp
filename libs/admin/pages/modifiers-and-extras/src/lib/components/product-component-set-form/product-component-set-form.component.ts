import { combineLatest, iif, Observable, of } from 'rxjs';
import { startWith, switchMap, take, tap } from 'rxjs/operators';

import {
  ChangeDetectionStrategy,
  ChangeDetectorRef,
  Component,
  Injector,
  OnDestroy,
  OnInit,
} from '@angular/core';
import { FormGroup, Validators } from '@angular/forms';
import { AbstractFormDialogComponent } from '@bgap/admin/shared/forms';
import {
  getProductComponentObject,
  getProductComponentOptions,
  SERVING_MODES,
} from '@bgap/admin/shared/utils';
import { chainsSelectors } from '@bgap/admin/store/chains';
import { loggedUserSelectors } from '@bgap/admin/store/logged-user';
import { ProductComponentCollectionService } from '@bgap/admin/store/product-components';
import * as CrudApi from '@bgap/crud-gql/api';
import { KeyValue, KeyValueObject, UpsertResponse } from '@bgap/shared/types';
import { cleanObject } from '@bgap/shared/utils';
import { UntilDestroy, untilDestroyed } from '@ngneat/until-destroy';
import { select } from '@ngrx/store';

import { productComponentSetTypeOptions } from '../../const';
import { ModifiersAndExtrasFormService } from '../../services/modifiers-and-extras-form.service';

@UntilDestroy()
@Component({
  changeDetection: ChangeDetectionStrategy.OnPush,
  selector: 'bgap-product-component-set-form',
  templateUrl: './product-component-set-form.component.html',
})
export class ProductComponentSetFormComponent
  extends AbstractFormDialogComponent
  implements OnInit, OnDestroy
{
  public componentForm?: FormGroup;
  public productComponentSet!: CrudApi.ProductComponentSet;
  public chainOptions$: Observable<KeyValue[]>;
  public typeOptions: KeyValue[] = [];
  public productComponentOptions: KeyValue[] = [];
  public productComponentObject: KeyValueObject = {};
  public eProductComponentSetType = CrudApi.ProductComponentSetType;
  public servingModes = SERVING_MODES;
  public editing = false;

  constructor(
    protected _injector: Injector,
    private _changeDetectorRef: ChangeDetectorRef,
    private _modifiersAndExtrasFormService: ModifiersAndExtrasFormService,
    private _productComponentCollectionService: ProductComponentCollectionService,
  ) {
    super(_injector);

    this.dialogForm =
      this._modifiersAndExtrasFormService.createProductComponentSetFormGroup();

    this.chainOptions$ = this._store.select(
      chainsSelectors.getAllChainOptions(),
    );

    this.componentForm = this._formBuilder.group({
      productComponentId: ['', [Validators.required]],
    });

    this.typeOptions = productComponentSetTypeOptions;
  }

  ngOnInit() {
    if (this.productComponentSet) {
      this.dialogForm?.patchValue(cleanObject(this.productComponentSet));
    } else {
      // Patch ChainId
      this._store
        .pipe(select(loggedUserSelectors.getSelectedChainId), take(1))
        .subscribe((selectedChainId: string | undefined | null): void => {
          if (selectedChainId) {
            this.dialogForm?.controls.chainId.patchValue(selectedChainId);
          }
        });
    }

    combineLatest([
      this._productComponentCollectionService.filteredEntities$,
      iif(
        () => typeof this.dialogForm !== 'undefined',
        this.dialogForm?.controls['items'].valueChanges.pipe(
          startWith(this.dialogForm?.value.items || []),
        ),
        of([]),
      ),
    ])
      .pipe(untilDestroyed(this))
      .subscribe(([productComponents, items]) => {
        this.productComponentOptions = getProductComponentOptions(
          productComponents,
          items,
        );
        this.productComponentObject =
          getProductComponentObject(productComponents);

        this._changeDetectorRef.markForCheck();
      });

    this._changeDetectorRef.detectChanges();
  }

  ngOnDestroy(): void {
    // untilDestroyed uses it.
  }

  public addComponentToList() {
    if (this.dialogForm && this.componentForm) {
      this._modifiersAndExtrasFormService.addComponentToList(
        this.dialogForm,
        this.componentForm,
      );
    }
  }

  public removeComponentFromList(productComponentId: string) {
    if (this.dialogForm) {
      this._modifiersAndExtrasFormService.removeComponentFromList(
        this.dialogForm,
        productComponentId,
      );
    }
  }

  public move(idx: number, change: number) {
    const componentIdsArr: string[] = this.dialogForm?.controls['items'].value;
    const itemId = componentIdsArr[idx];

    componentIdsArr.splice(idx, 1);
    componentIdsArr.splice(idx + change, 0, itemId);
  }

  public submit() {
    if (this.dialogForm?.valid) {
      this.setWorking$(true)
        .pipe(
          switchMap(() =>
            this._modifiersAndExtrasFormService.saveComponentSetForm$(
              {
                ...this.dialogForm?.value,
                maxSelection:
                  this.dialogForm?.value.type ===
                  CrudApi.ProductComponentSetType.modifier
                    ? null
                    : this.dialogForm?.value.maxSelection,
                dirty: this.productComponentSet?.dirty ? false : undefined,
              },
              this.productComponentSet?.id,
            ),
          ),
          tap(() => this.setWorking$(false)),
        )
        .subscribe((response: UpsertResponse<unknown>) => {
          this._toasterService.showSimpleSuccess(response.type);

          this.close();
        });
    }
  }
}
