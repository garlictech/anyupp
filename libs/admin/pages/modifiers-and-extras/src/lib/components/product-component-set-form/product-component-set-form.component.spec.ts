import { CrudSdkService } from 'libs/admin/shared/data-access/sdk/src';
import { switchMap, tap } from 'rxjs/operators';

import { ComponentFixture, TestBed, waitForAsync } from '@angular/core/testing';
import { FormsModule, ReactiveFormsModule } from '@angular/forms';
import { Auth } from '@aws-amplify/auth';
import { DataService } from '@bgap/admin/shared/data-access/data';
import { CHAIN_ID_01, initialStateFixture } from '@bgap/admin/shared/fixtures';
import { AdminSharedFormsModule } from '@bgap/admin/shared/forms';
import { AdminSharedPipesModule } from '@bgap/admin/shared/pipes';
import * as CrudApi from '@bgap/crud-gql/api';
import {
  getCognitoUsername,
  testAdminUserPassword,
} from '@bgap/shared/fixtures';
import {
  defaultServingMode,
  EProductComponentSetType,
} from '@bgap/shared/types';
import {
  NbCardModule,
  NbDialogModule,
  NbDialogRef,
  NbIconModule,
  NbListModule,
  NbThemeModule,
  NbToastrService,
} from '@nebular/theme';
import { provideMockStore } from '@ngrx/store/testing';
import { TranslateModule } from '@ngx-translate/core';

import { ProductComponentSetFormComponent } from './product-component-set-form.component';

const mockProductComponentSetData = {
  chainId: CHAIN_ID_01,
  type: EProductComponentSetType.EXTRAS,
  maxSelection: 1,
  name: {
    en: 'mySet',
  },
  description: 'mock desc',
  supportedServingModes: [defaultServingMode],
  items: ['component_id'],
};

describe('ProductComponentSetFormComponent', (): void => {
  let component: ProductComponentSetFormComponent;
  let fixture: ComponentFixture<ProductComponentSetFormComponent>;
  let crudSdk: CrudSdkService;

  beforeAll(async () => {
    Auth.configure({
      ...CrudApi.awsConfig,
      authenticationFlowType: 'USER_PASSWORD_AUTH',
      aws_appsync_authenticationType: 'AMAZON_COGNITO_USER_POOLS',
    });

    await Auth.signIn(getCognitoUsername('monad'), testAdminUserPassword);
  });

  beforeEach(
    waitForAsync((): void => {
      TestBed.configureTestingModule({
        declarations: [ProductComponentSetFormComponent],
        imports: [
          TranslateModule.forRoot(),
          NbThemeModule.forRoot(),
          NbDialogModule.forRoot(),
          NbCardModule,
          NbIconModule,
          NbListModule,
          AdminSharedPipesModule,
          AdminSharedFormsModule,
          FormsModule,
          ReactiveFormsModule,
        ],
        providers: [
          provideMockStore(initialStateFixture),
          {
            provide: NbDialogRef,
            useValue: { close: jest.fn() },
          },
          {
            provide: NbToastrService,
            useValue: { show: jest.fn() },
          },
          {
            provide: DataService,
            useValue: {},
          },
          CrudSdkService,
        ],
      }).compileComponents();

      fixture = TestBed.createComponent(ProductComponentSetFormComponent);
      component = fixture.componentInstance;
      crudSdk = new CrudSdkService(component['_store']);
    }),
  );

  it('should create', (): void => {
    expect(component).toBeTruthy();
  });

  it('should create empty form', (): void => {
    component.ngOnInit();

    expect(component.dialogForm.value).toMatchSnapshot();
  });

  it('should save new component set data', done => {
    // Inherited from a chain product
    component.editing = false;
    component.ngOnInit();

    // "Fill" the new values
    component.dialogForm.patchValue(mockProductComponentSetData);

    component['_save']()
      .pipe(
        switchMap(saveResponse => {
          expect(component.dialogForm.valid).toBeTruthy();
          expect(<CrudApi.ProductComponentSet>saveResponse).toMatchObject(
            mockProductComponentSetData,
          );

          // Delete test data
          return crudSdk.sdk.DeleteProductComponentSet({
            input: { id: (<CrudApi.ProductComponentSet>saveResponse).id },
          });
        }),
      )
      .subscribe({
        next() {
          done();
        },
        error(err) {
          console.error('ERR', err);
        },
      });
  }, 15000);

  it('should save updated component set data', done => {
    // Inherited from a group product
    component.productComponentSet = {
      ...mockProductComponentSetData,
      id: 'mock_id',
      createdAt: '2021-10-01T01:54:11.843Z',
      updatedAt: '2021-10-01T01:54:11.843Z',
    };
    component.editing = true;
    component.ngOnInit();

    // Change the form
    component.dialogForm.patchValue({
      description: 'modded',
      supportedServingModes: [CrudApi.ServingMode.takeaway],
    });

    // Create a base product and update it with the form above
    crudSdk.sdk
      .CreateProductComponentSet({ input: mockProductComponentSetData })
      .pipe(
        tap(createdData => {
          component.productComponentSet!.id = (<CrudApi.ProductComponentSet>(
            createdData
          )).id;
        }),
        // Save the modified form
        switchMap(() => component['_save']()),
        switchMap(saveResponse => {
          expect(component.dialogForm.valid).toBeTruthy();
          expect(
            (<CrudApi.ProductComponentSet>saveResponse).description,
          ).toEqual('modded');
          expect(
            (<CrudApi.ProductComponentSet>saveResponse).supportedServingModes,
          ).toEqual([CrudApi.ServingMode.takeaway]);

          // Delete test data
          return crudSdk.sdk.DeleteProductComponentSet({
            input: { id: (<CrudApi.ProductComponentSet>saveResponse).id },
          });
        }),
      )
      .subscribe({
        next() {
          done();
        },
        error(err) {
          console.log('ERR', err);
        },
      });
  }, 15000);
});
