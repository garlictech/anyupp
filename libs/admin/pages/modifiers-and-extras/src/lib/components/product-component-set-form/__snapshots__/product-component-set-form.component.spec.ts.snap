// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ProductComponentSetFormComponent should create empty form 1`] = `
Object {
  "chainId": "",
  "description": "",
  "items": Array [],
  "maxSelection": "",
  "name": Object {
    "de": "",
    "en": "",
    "hu": "",
  },
  "supportedServingModes": Array [
    "inPlace",
  ],
  "type": "",
}
`;
